sys/drv/hidm.c

```diff
@@ -0,0 +1,186 @@
+typedef struct hidm_state {
+  thread_t *thread;     /* input packets gathering thread */
+  evdev_dev_t *evdev;   /* evdev device structure */
+  uint8_t prev_buttons; /* the most recent buttons state */
+} hidm_state_t;
+
+/*
+ * Evdev handling functions.
+ */
+
+static void hidm_init_evdev(device_t *dev) {
+  hidm_state_t *hidm = dev->state;
+  evdev_dev_t *evdev = evdev_alloc();
+
+  /* Set the basic evdev parameters. */
+  usb_device_t *udev = usb_device_of(dev);
+  evdev_set_name(evdev, "HID mouse");
+  evdev_set_id(evdev, BUS_USB, udev->vendor_id, udev->product_id, 0);
+
+  /* `EV_SYN` is required for every evdev device. */
+  evdev_support_event(evdev, EV_SYN);
+
+  /* Mouse buttons. */
+  evdev_support_event(evdev, EV_KEY);
+  evdev_support_all_mouse_buttons(evdev);
+
+  /* Mouse movement. */
+  evdev_support_event(evdev, EV_REL);
+  evdev_support_rel(evdev, REL_X);
+  evdev_support_rel(evdev, REL_Y);
+
+  /* Bind `evdev` to `hidm` for future references. */
+  hidm->evdev = evdev;
+
+  /* Finally, register the device in the file system. */
+  evdev_register(evdev);
+}
+
+/*
+ * Report processing functions.
+ */
+
+static void hidm_process_buttons(hidm_state_t *hidm, uint8_t buttons) {
+  uint8_t prev_buttons = hidm->prev_buttons;
+
+  for (size_t i = 0; i < HIDM_NBUTTONS; i++) {
+    uint8_t prev = prev_buttons & (1 << i);
+    uint8_t cur = buttons & (1 << i);
+
+    if ((cur == prev) && !cur)
+      continue;
+
+    uint16_t keycode = evdev_hid2btn(i);
+    evdev_push_key(hidm->evdev, keycode, cur);
+    evdev_sync(hidm->evdev);
+  }
+}
+
+static void hidm_process_in_report(hidm_state_t *hidm,
+                                   hidm_in_report_t *report) {
+  /* Hnadle buttons. */
+  hidm_process_buttons(hidm, report->buttons);
+
+  /* Handle the X-axis. */
+  int8_t x = report->x;
+  if (x) {
+    evdev_push_event(hidm->evdev, EV_REL, REL_X, x);
+    evdev_sync(hidm->evdev);
+  }
+
+  /* Handle the Y-axis. */
+  int8_t y = report->y;
+  if (y) {
+    evdev_push_event(hidm->evdev, EV_REL, REL_Y, y);
+    evdev_sync(hidm->evdev);
+  }
+}
+
+/* A thread which gathers input reports, processes and converts them
+ * to evdev-compatible events and then hands them over to the evdev layer. */
+static void hidm_thread(void *arg) {
+  device_t *dev = arg;
+  hidm_state_t *hidm = dev->state;
+  hidm_in_report_t report;
+
+  /* Obtain a USB buf. */
+  usb_buf_t *buf = usb_buf_alloc();
+
+  /* Register an interrupt input transfer. */
+  usb_data_transfer(dev, buf, &report, sizeof(hidm_in_report_t),
+                    USB_TFR_INTERRUPT, USB_DIR_INPUT);
+
+  for (;;) {
+    /* Wait for the data to arrvie. */
+    int error = usb_buf_wait(buf);
+    if (!error) {
+      hidm_process_in_report(hidm, &report);
+      hidm->prev_buttons = report.buttons;
+      continue;
+    }
+
+    /* Recover and reenable the transfer. */
+    if ((error = usb_unhalt_endpt(dev, USB_TFR_INTERRUPT, USB_DIR_INPUT)))
+      panic("Unable to unhalt an endpoint");
```

Comment:
This is a bit worrying. Can we react in any reasonable way here?

Reply:
FTTB, I left a command pointing out that we should implement evdev device unregistering.